cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -41,11 +41,122 @@ static int sim_ending = 0;
 
 #include "simulatormodule.h"
 #include <cocotb_utils.h>     // COCOTB_UNUSED
+#include <type_traits>
+
 
 struct module_state {
     PyObject *error;
 };
 
+
+/* define the extension types as templates */
+namespace {
+    template<typename gpi_hdl>
+    struct gpi_hdl_Object {
```

Comment:
I'm not a fan of these mixed underscore_camelCase names. We probably have to die one death, but internal consistency would go for gpi_hdl_object and gpi_hdl_new.

Reply:
`_Object` and `_New` is to invoke similarities to `PyTupleObject` and `PyTuple_New`. I could go for `GpiHdlObject` instead if you'd prefer, but then that bears less resemblance to the types from `gpi.h`.

Reply:
That's what I mean dying one death: `gpi_hdl` isn't named the way you'd expect Python objects to be named (PythonObject), which makes the `_New` suffix look strange. 

In our C/C++ code base, we go for lower_snake_case, so gpi_hdl_object and gpi_hdl_new would fit that scheme. GpiHdlObject isn't consistent with the rest of our C/C++ code base, so that's not really an option IMO. 

Up to you: either keep the current "compromise solution" with "gpi_hdl_Object", or go for "gpi_hdl_object" to fully match our style, giving up the similarity to the Python code base.

Reply:
> GpiHdlObject isn't consistent with the rest of our C/C++ code base

Not sure what you mean, in `GpiCommon.cpp` we already have `GpiObjHdl`.

Reply:
We just have this C interface that renames all our `PascalCaseClasses::with_snake_methods` to `gpi_class_snake_method`, so have to juggle multiple naming conventions.

Reply:
`PyGpiHdlObject` might also be sensible, although I'm not quite sure if we're supposed to use the `Py` prefix.